File: java/tema_3/src/main/java/ro/mpp2024/concertpro/dao/repository/spectacle_repository/SpectacleDbRepository.java
package ro.mpp2024.concertpro.dao.repository.spectacle_repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.concertpro.dao.model.Spectacle;
import ro.mpp2024.concertpro.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class SpectacleDbRepository implements ISpectacleRepository {
    private final JdbcUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();

    public SpectacleDbRepository(Properties props) {
        logger.info("Initializing SpectacleDbRepository with properties: {} ", props);
        this.jdbcUtils = new JdbcUtils(props);
    }

    @Override
    public Spectacle findOne(Long spectacleIdN) {
        logger.traceEntry();
        Connection connection = this.jdbcUtils.getConnection();
        Spectacle spectacle = new Spectacle();

        try(PreparedStatement preparedStatement = connection.prepareStatement("SELECT * FROM spectacles WHERE " +
                "spectacleId = ?")) {
            preparedStatement.setLong(1, spectacleIdN);

            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                while(resultSet.next()) {
                    Long spectacleId = resultSet.getLong("spectacleId");
                    String artistName = resultSet.getString("artistName");
                    LocalDate spectacleDate = resultSet.getDate("spectacleDate").toLocalDate();
                    String spectaclePlace = resultSet.getString("spectaclePlace");
                    Long seatsAvailable = resultSet.getLong("seatsAvailable");
                    Long seatsSold = resultSet.getLong("seatsSold");

                    spectacle.setId(spectacleId);
                    spectacle.setArtistName(artistName);
                    spectacle.setSpectacleDate(spectacleDate);
                    spectacle.setSpectaclePlace(spectaclePlace);
                    spectacle.setSeatsAvailable(seatsAvailable);
                    spectacle.setSeatsSold(seatsSold);
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
            System.err.println("Error DB: " + exception);
        }

        logger.traceExit();
        return spectacle;
    }

    @Override
    public Iterable<Spectacle> findAll() {
        logger.traceEntry("findAll Spectacles task");
        Connection connection = this.jdbcUtils.getConnection();
        List<Spectacle> spectacles = new ArrayList<>();

        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from spectacles")) {
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Long spectacleId = resultSet.getLong("spectacleId");
                    String artistName = resultSet.getString("artistName");
                    LocalDate spectacleDate = resultSet.getDate("spectacleDate").toLocalDate();
                    String spectaclePlace = resultSet.getString("spectaclePlace");
                    Long seatsAvailable = resultSet.getLong("seatsAvailable");
                    Long seatsSold = resultSet.getLong("seatsSold");

                    Spectacle spectacle = new Spectacle(artistName, spectacleDate, spectaclePlace, seatsAvailable, seatsSold);
                    spectacle.setId(spectacleId);
                    spectacles.add(spectacle);
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
            System.err.println("Error DB: " + exception);
        }

        logger.traceExit();
        return spectacles;
    }

    @Override
    public void save(Spectacle spectacle) {
    }

    @Override
    public void update(Spectacle spectacle, Long spectacleId) {
        logger.traceEntry("update task {} ", spectacle);
        Connection connection = this.jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("update spectacles set " +
                "artistName = ?, spectacleDate = ?, spectaclePlace = ?, seatsAvailable = ?, seatsSold = ? " +
                "where spectacleId = ?")) {
            preparedStatement.setString(1, spectacle.getArtistName());
            preparedStatement.setString(2, spectacle.getSpectacleDate().toString());
            preparedStatement.setString(3, spectacle.getSpectaclePlace());
            preparedStatement.setLong(4, spectacle.getSeatsAvailable());
            preparedStatement.setLong(5, spectacle.getSeatsSold());
            preparedStatement.setLong(6, spectacleId);

            int result = preparedStatement.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch (SQLException exception) {
            logger.error(exception);
            System.err.println("Error DB: " + exception);
        }

        logger.traceExit();
    }

    @Override
    public void delete(Long spectacleId) {
    }

    @Override
    public Iterable<Spectacle> getAllByDate(LocalDate date) {
        logger.traceEntry("findSpectaclesByDate task");
        Connection connection = this.jdbcUtils.getConnection();
        List<Spectacle> spectacles = new ArrayList<>();

        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from spectacles where spectacleDate = ?")) {
            preparedStatement.setString(1, date.toString());

            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Long spectacleId = resultSet.getLong("spectacleId");
                    String artistName = resultSet.getString("artistName");
                    LocalDate spectacleDate = resultSet.getDate("spectacleDate").toLocalDate();
                    String spectaclePlace = resultSet.getString("spectaclePlace");
                    Long seatsAvailable = resultSet.getLong("seatsAvailable");
                    Long seatsSold = resultSet.getLong("seatsSold");

                    Spectacle spectacle = new Spectacle(artistName, spectacleDate, spectaclePlace, seatsAvailable, seatsSold);
                    spectacle.setId(spectacleId);
                    spectacles.add(spectacle);
                }
            }
        } catch (SQLException exception) {
            logger.error(exception);
            System.err.println("Error DB: " + exception);
        }

        logger.traceExit();
        return spectacles;
    }
}

File: c#/tema_5/ConcertPro/Dao/Repository/SpectacleRepository/SpectacleDbRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using ConcertPro.Dao.Model;
using log4net;

namespace ConcertPro.Dao.Repository.SpectacleRepository
{
    public class SpectacleDbRepository : ISpectacleRepository
    {
        private static readonly ILog Log = LogManager.GetLogger("SpectacleDbRepository");
        private readonly IDictionary<string, string> _properties;

        public SpectacleDbRepository(IDictionary<string, string> properties)
        {
            Log.Info("Initializing SpectacleDbRepository with properties: " + properties);
            _properties = properties;
        }
        
        public Spectacle FindOne(long id)
        {
            Log.InfoFormat("Entering FindOneSpectacle with value {0}", id);
            IDbConnection connection = DbUtils.DbUtils.GetConnection(_properties);

            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "select * from spectacles where spectacleId = @spectacleId";
                IDbDataParameter spectacleIdParameter = command.CreateParameter();
                spectacleIdParameter.ParameterName = "@spectacleId";
                spectacleIdParameter.Value = id;
                command.Parameters.Add(spectacleIdParameter);

                using (IDataReader dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        long spectacleId = dataReader.GetInt64(0);
                        string artistName = dataReader.GetString(1);
                        DateTime spectacleDate = Convert.ToDateTime(dataReader.GetDateTime(2));
                        string spectaclePlace = dataReader.GetString(3);
                        long seatsAvailable = dataReader.GetInt64(4);
                        long seatsSold = dataReader.GetInt64(5);

                        Spectacle spectacle = new Spectacle(artistName, spectacleDate, spectaclePlace, seatsAvailable,
                            seatsSold);
                        spectacle.Id = spectacleId;
                        
                        Log.InfoFormat("Existing findOneSpectacle with value {0}", spectacle);
                        return spectacle;
                    }
                }
            }
            
            Log.InfoFormat("Existing findOneSpectacle with value {0}", null);
            return null;
        }

        public IEnumerable<Spectacle> FindAll()
        {
            Log.InfoFormat("find all spectacles task");
            IDbConnection connection = DbUtils.DbUtils.GetConnection(_properties);
            IList<Spectacle> spectacles = new List<Spectacle>();

            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "select * from spectacles";

                using (IDataReader dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        long spectacleId = dataReader.GetInt64(0);
                        string artistName = dataReader.GetString(1);
                        DateTime spectacleDate = dataReader.GetDateTime(2);
                        string spectaclePlace = dataReader.GetString(3);
                        long seatsAvailable = dataReader.GetInt64(4);
                        long seatsSold = dataReader.GetInt64(5);

                        Spectacle spectacle = new Spectacle(artistName, spectacleDate, spectaclePlace, seatsAvailable,
                            seatsSold);
                        spectacle.Id = spectacleId;
                        spectacles.Add(spectacle);
                    }
                }
            }

            return spectacles;
        }

        public void Save(Spectacle entity)
        {
            throw new NotImplementedException();
        }

        public void Update(Spectacle entity, long id)
        {
            Log.InfoFormat("Update Spectacle Task {0}", entity);
            IDbConnection connection = DbUtils.DbUtils.GetConnection(_properties);

            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "update spectacles set artistName = @artistName, " +
                                      "spectacleDate = @spectacleDate, spectaclePlace = @spectaclePlace, " +
                                      "seatsAvailable = @seatsAvailable, seatsSold = @seatsSold where " +
                                      "spectacleId = @spectacleId";
                
                IDbDataParameter artistNameParam = command.CreateParameter();
                artistNameParam.ParameterName = "@artistName";
                artistNameParam.Value = entity.ArtistName;
                command.Parameters.Add(artistNameParam);
                
                IDbDataParameter spectacleDateParam = command.CreateParameter();
                spectacleDateParam.ParameterName = "@spectacleDate";
                spectacleDateParam.Value = entity.SpectacleDate;
                command.Parameters.Add(spectacleDateParam);
                
                IDbDataParameter spectaclePlaceParameter = command.CreateParameter();
                spectaclePlaceParameter.ParameterName = "@spectaclePlace";
                spectaclePlaceParameter.Value = entity.SpectaclePlace;
                command.Parameters.Add(spectaclePlaceParameter);
                
                IDbDataParameter seatsAvailableParameter = command.CreateParameter();
                seatsAvailableParameter.ParameterName = "@seatsAvailable";
                seatsAvailableParameter.Value = entity.SeatsAvailable;
                command.Parameters.Add(seatsAvailableParameter);
                
                IDbDataParameter seatsSoldParameter = command.CreateParameter();
                seatsSoldParameter.ParameterName = "@seatsSold";
                seatsSoldParameter.Value = entity.SeatsSold;
                command.Parameters.Add(seatsSoldParameter);
                
                IDbDataParameter spectacleIdParameter = command.CreateParameter();
                spectacleIdParameter.ParameterName = "@spectacleId";
                spectacleIdParameter.Value = id;
                command.Parameters.Add(spectacleIdParameter);

                int result = command.ExecuteNonQuery();
                if (result == 0)
                {
                    Log.Error("Error DB: No Spectacle Updated");
                }
                
                Log.InfoFormat("Saved {0} instances", result);
            }
        }

        public void Delete(long id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<Spectacle> GetAllByDate(DateTime date)
        {
            Log.InfoFormat("findAllSpectaclesByDate task");
            IDbConnection connection = DbUtils.DbUtils.GetConnection(_properties);
            IList<Spectacle> spectacles = new List<Spectacle>();

            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "select * from spectacles where spectacleDate = @spectacleDate";
                IDbDataParameter spectacleDateParameter = command.CreateParameter();
                spectacleDateParameter.ParameterName = "@spectacleDate";
                spectacleDateParameter.Value = date;
                command.Parameters.Add(spectacleDateParameter);

                using (IDataReader dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        long spectacleId = dataReader.GetInt64(0);
                        string artistName = dataReader.GetString(1);
                        DateTime spectacleDate = dataReader.GetDateTime(2);
                        string spectaclePlace = dataReader.GetString(3);
                        long seatsAvailable = dataReader.GetInt64(4);
                        long seatsSold = dataReader.GetInt64(5);

                        Spectacle spectacle = new Spectacle(artistName, spectacleDate, spectaclePlace, seatsAvailable,
                            seatsSold);
                        spectacle.Id = spectacleId;
                        spectacles.Add(spectacle);
                    }
                }
            }

            return spectacles;
        }
    }
}
